pkg/sunrise/verify.go

```diff
@@ -0,0 +1,97 @@
+package sunrise
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+const (
+	nonceLength    = 64
+	keyLength      = 64
+	minTokenLength = 16 + nonceLength + 1
+	maxTokenLength = 16 + nonceLength + binary.MaxVarintLen64
+)
+
+type Token struct {
+	EnvelopeID uuid.UUID
+	Expiration time.Time
+	nonce      []byte
+}
+
+func NewToken(envelopeID uuid.UUID, expiration time.Time) *Token {
+	token := &Token{
+		EnvelopeID: envelopeID,
+		Expiration: expiration,
+		nonce:      make([]byte, nonceLength),
+	}
+
+	if _, err := rand.Read(token.nonce); err != nil {
+		panic(fmt.Errorf("no crypto random generator available: %w", err))
+	}
+
+	return token
+}
+
+func (t *Token) MarshalBinary() ([]byte, error) {
+	if err := t.Validate(); err != nil {
+		return nil, err
+	}
+
+	data := make([]byte, maxTokenLength)
+	copy(data[:16], t.EnvelopeID[:])
+
+	i := binary.PutVarint(data[16:], t.Expiration.UnixNano())
+	copy(data[16+i:], t.nonce)
+
+	return data[:16+i+nonceLength], nil
+}
```

Comment:
I'm doing manual marshaling and unmarshaling rather than using something like `msgpack` to make the binary data as compact as possible so we don't have a crazy long token at the end.